src/minisketch/src/sketch_impl.h

```diff
@@ -0,0 +1,433 @@
+                b_inv = field.Inv(b);
+                b_have_inv = true;
+            }
+            bool swap = 2 * (current.size() - 1) <= n;
+            if (swap) {
+                if (prev.size() + x - 1 > max_degree) return {}; // We'd exceed maximum degree
+                tmp = current;
+                current.resize(prev.size() + x);
+            }
+            typename F::Multiplier mul(field, field.Mul(discrepancy, b_inv));
+            for (size_t i = 0; i < prev.size(); ++i) current[i + x] ^= mul(prev[i]);
+            if (swap) {
+                std::swap(prev, tmp);
+                b = discrepancy;
+                b_have_inv = false;
+            }
+        }
+    }
+    CHECK_RETURN(current.size() && current.back() != 0, {});
+    return current;
+}
+
+template<typename F>
+std::vector<typename F::Elem> ReconstructAllSyndromes(const std::vector<typename F::Elem>& odd_syndromes, const F& field) {
+    std::vector<typename F::Elem> all_syndromes;
+    all_syndromes.resize(odd_syndromes.size() * 2);
+    for (size_t i = 0; i < odd_syndromes.size(); ++i) {
+        all_syndromes[i * 2] = odd_syndromes[i];
+        all_syndromes[i * 2 + 1] = field.Sqr(all_syndromes[i]);
+    }
+    return all_syndromes;
+}
+
+template<typename F>
+void AddToOddSyndromes(std::vector<typename F::Elem>& osyndromes, typename F::Elem data, const F& field) {
+    auto sqr = field.Sqr(data);
+    typename F::Multiplier mul(field, sqr);
+    for (auto& osyndrome : osyndromes) {
+        osyndrome ^= data;
+        data = mul(data);
+    }
+}
+
+template<typename F>
+std::vector<typename F::Elem> FullDecode(const std::vector<typename F::Elem>& osyndromes, const F& field) {
+    auto asyndromes = ReconstructAllSyndromes<typename F::Elem>(osyndromes, field);
+    auto poly = BerlekampMassey(asyndromes, field);
+    std::reverse(poly.begin(), poly.end());
+    return FindRoots(poly, field);
+}
+
+template<typename F>
+class SketchImpl final : public Sketch
+{
+    const F m_field;
+    std::vector<typename F::Elem> m_syndromes;
+    typename F::Elem m_basis;
+
+public:
+    template<typename... Args>
+    SketchImpl(int implementation, int bits, const Args&... args) : Sketch(implementation, bits), m_field(args...) {
+        std::random_device rng;
+        std::uniform_int_distribution<uint64_t> dist;
+        m_basis = m_field.FromSeed(dist(rng));
+    }
+
+    size_t Syndromes() const override { return m_syndromes.size(); }
+    void Init(int count) override { m_syndromes.assign(count, 0); }
+
+    void Add(uint64_t val) override
+    {
+        auto elem = m_field.FromUint64(val);
+        AddToOddSyndromes(m_syndromes, elem, m_field);
+    }
+
+    void Serialize(unsigned char* ptr) const override
+    {
+        BitWriter writer(ptr);
+        for (const auto& val : m_syndromes) {
+            m_field.Serialize(writer, val);
+        }
+        writer.Flush();
+    }
+
+    void Deserialize(const unsigned char* ptr) override
```

Comment:
May be helpful to add a fuzz harness for sketch deserialization. A large number of security bugs come from deserializing untrusted data.